compiler/llvm/llvmDebug.cpp

```diff
@@ -380,126 +522,308 @@ llvm::DIType* debug_data::construct_type_for_pointer(llvm::Type* ty, Type* type)
   } // end of else (type==type->getType)
+  if (developer || fVerify) {
+    INT_FATAL("Unhandled debug info generation for pointer type: "
+              "%s\n\ttype->astTag=%i\n\tllvmImplType->getTypeID()=%i\n",
+              type->symbol->name, type->astTag, ty->getTypeID());
+  }
   return nullptr;
 }
 
-llvm::DIType* debug_data::construct_type_for_special_cases(llvm::Type* ty, Type* type) {
+llvm::DIType* DebugData::constructTypeFromChplType(llvm::Type* ty, Type* type) {
 
   GenInfo* info = gGenInfo;
   const llvm::DataLayout& layout = info->module->getDataLayout();
 
   const char* name = type->symbol->name;
-  ModuleSymbol* defModule = type->symbol->getModule();
-  const char* defFile = type->symbol->fname();
-  int defLine = type->symbol->linenum();
-
-  auto dibuilder = defModule->llvmDIBuilder;
+  DefinitionInfo defInfo(this, type->symbol);
+  auto dibuilder = type->symbol->getModule()->llvmDIBuilder;
 
 
-  if (type == dtObject) {
-    // wrap_in_pointer_if_needed(llvm::cast<llvm::DIType>(N), type);
+  if (type->isRef()) {
+    auto valType = type->getValType();
+    auto diType = valType ? getType(valType) : nullptr;
+    if (!diType) {
+      if (developer || fVerify) {
+        INT_FATAL("Unable to find DIType for ref type %s ", type->symbol->name);
+      }
+      return nullptr;
+    }
+    auto N = dibuilder->createReferenceType(llvm::dwarf::DW_TAG_reference_type, diType);
+    N = dibuilder->createTypedef(N, name, nullptr, 0, nullptr);
+    type->symbol->llvmDIType = N;
+    return N;
+  } else if (type == dtObject) {
     llvm::SmallVector<llvm::Metadata *, 1> EltTys;
     llvm::Type* cidType = info->lvt->getType("chpl__class_id");
     auto cidDITy = dibuilder->createMemberType(
-      get_module_scope(defModule),
+      defInfo.maybeScope(),
       "cid",
-      get_file(defModule, defFile),
-      defLine,
+      defInfo.file(),
+      defInfo.line(),
       layout.getTypeSizeInBits(cidType),
       8*layout.getABITypeAlign(cidType).value(),
       0, /* offset, assume its zero */
       llvm::DINode::FlagZero,
-      get_type(CLASS_ID_TYPE)
+      getType(CLASS_ID_TYPE)
     );
     EltTys.push_back(cidDITy);
     // since dtOject has a single field, we can directly assume the size and alignent to
     // be the same as its single field
     llvm::DIType* N = dibuilder->createStructType(
-      get_module_scope(defModule), /* Scope */
-      name, /* Name */
-      get_file(defModule, defFile), /* File */
-      defLine, /* LineNumber */
+      defInfo.maybeScope(),
+      name,
+      defInfo.file(),
+      defInfo.line(),
       layout.getTypeSizeInBits(cidType), /* SizeInBits */
       8*layout.getABITypeAlign(cidType).value(), /* AlignInBits */
       llvm::DINode::FlagZero, /* Flags */
       nullptr,
       dibuilder->getOrCreateArray(EltTys) /* Elements */
     );
-    N = wrap_in_pointer_if_needed(llvm::cast<llvm::DIType>(N), type);
+    N = maybeWrapTypeInPointer(N, type);
```

Comment:
is this the section you were thinking about having a callback table for? might be good to put a TODO for that